File: frontend/src/services/doctor.service.ts
import * as api from '@/api/doctor.api';
import type { Doctor, DoctorCreate, DoctorUpdate } from '@/types/medical/doctor.model';
import type { Token } from '@/types/user/token';

import { withAxiosHandler } from './commons';

/**
 * Service to create a new doctor in the API service.
 * @param doctor - DoctorCreate
 * @param token - Token['accessToken']
 * @returns a Promise that resolves to a Doctor.
 */
export const createDoctor: (
  doctor: DoctorCreate,
  token: Token['accessToken']
) => Promise<Doctor> = withAxiosHandler(async (doctor, token) =>
  api.create(doctor, token)
);

/**
 * Service to update a doctor in the API service.
 * @param doctor - DoctorUpdate
 * @param token - Token['accessToken']
 * @returns a Promise that resolves to a Doctor.
 */
export const updateDoctor: (
  dni: Doctor['dni'],
  doctor: DoctorUpdate,
  token: Token['accessToken']
) => Promise<Doctor> = withAxiosHandler((dni, doctor, token) =>
  api.update(dni, doctor, token)
);

File: backend/app/services/medical/doctor_service.py
from sqlalchemy.orm import Session
from sqlalchemy.sql import true

from app.models.medical import Doctor
from app.models.user import User
from app.schemas.medical.doctor import DoctorCreate, DoctorUpdate
from app.services.common.base import BaseService


class DoctorService(BaseService[Doctor, DoctorCreate, DoctorUpdate]):
    '''Service that provides CRUD operations for a doctor model.

    Args:
        BaseService ([Doctor, DoctorCreate, DoctorUpdate]): Models and schemas.
    '''

    @classmethod
    def get_service(cls, database: Session):
        return cls(model=Doctor, database=database)

    def get_all(self, *, skip: int = 0, limit: int = 50) -> list[Doctor]:
        return (
            self.database.query(Doctor)
            .join(User)
            .filter(User.is_active == true())
            .slice(skip, limit)
            .all()
        )

    def is_active(self, doctor_id: int) -> bool:
        '''Checks if a doctor is active.

        Args:
            doctor_id (int): Doctor's id.

        Returns:
            bool: True if doctor is active, False otherwise.
        '''

        return self.database.query(User.is_active).filter(User.dni == doctor_id).scalar()
